lib/publiccloud/ssh_interactive.pm

```diff
@@ -1,67 +1,98 @@
 
 # Summary: Class with helpers related to SSH Interactive mode
 #
-# Maintainer: Pavel Dostal <pdostal@suse.cz>
+# Maintainer: qa-c@suse.de
 
 package publiccloud::ssh_interactive;
+use testapi qw(is_serial_terminal);
 use base Exporter;
 use testapi;
 use Utils::Backends qw(set_sshserial_dev unset_sshserial_dev);
+use utils qw(type_string_very_slow);
 use strict;
 use warnings;
+use distribution;
 
 our @EXPORT = qw(ssh_interactive_tunnel ssh_interactive_leave select_host_console);
 
+sub establish_tunnel_console {
+    # Helper call to activate a console and establish the ssh connection therein
+    my $console = $_[0];
+
+    select_console("$console");
+    # Note: Don't use script_run here! The serial terminal is set to /dev/sshserial, so every script_run will time out
+    type_string("\n~.\n", max_interval => 1);    # ensure no previous ssh connection is present
+    enter_cmd("clear");
+    enter_cmd('ssh -t sut');
+    # give the ssh connection some time to settle
+    sleep 5;
+}
+
 sub ssh_interactive_tunnel {
     # Establish the ssh interarctive tunnel to the publiccloud instance.
     # Optional arguments: 'force => 1' - reestablish tunnel, also if already established.
 
     my $instance = shift;
-    my %args = testapi::compat_args({cmd => undef}, ['cmd'], @_);
+    my %args = testapi::compat_args({force => 0, reconnect => 0}, ['force', 'reconnect'], @_);
     $args{force} //= 0;
+    $args{reconnect} //= 0;
 
     # $serialdev should be always set from os-autoinst/testapi.pm
     die '$serialdev is not set' if (!defined $serialdev || $serialdev eq '');
     return if ($args{force} != 1 && get_var('_SSH_TUNNELS_INITIALIZED', 0) == 1);
 
+    my $prev_console = current_console();
+    select_console('tunnel-console') unless ($prev_console =~ /tunnel-console/);
+
     # Prepare the environment for the SSH tunnel
     my $upload_port = get_required_var('QEMUPORT') + 1;
     my $upload_host = testapi::host_ip();
 
-    $instance->ssh_script_run(
-        # Create /dev/sshserial fifo on remote and tail|tee it to /dev/$serialdev on local
-        #   timeout => switches to script_run instead of script_output to be used so the test will not wait for the command to end
-        #   tunnel the worker port (for downloading from data/ and uploading assets / logs
-        cmd => "'rm -rf /dev/sshserial; mkfifo -m a=rwx /dev/sshserial; tail -fn +1 /dev/sshserial' 2>&1 | tee /dev/$serialdev; clear",
-        timeout => 0,
-        no_quote => 1,
-        ssh_opts => "-yt -R $upload_port:$upload_host:$upload_port",
-        username => 'root',
-    );
-    sleep 3;
-    save_screenshot;
+    # Pipe the output of the device fifo to the local serial terminal
+    background_script_run("while true; do ssh sut -R '$upload_port:$upload_host:$upload_port' 'rm -f /dev/sshserial && mkfifo -m a=rwx /dev/sshserial && tail -fn +1 /dev/sshserial' 2>&1 >/dev/$serialdev; sleep 5; done");
+    # give the ssh connection some time to settle
+    sleep 10;
```

Comment:
```suggestion
```
Why this is needed? It was not before. Let's not use `sleep` unless absolutely needed.

Reply:
As stated in the comment above, this is helpful to let ssh establish the connection to the guest system.
If we don't do this, it might happen that the following commands are being typed in too hastily.

Do you have a better idea on how to synchronize this? We need to wait for the ssh connection to be established and I didn't came up with an elegant way of doing this.

Reply:
Just run `ssh sut true` so you're sure that after such command the SSH is initiated.

Reply:
Or `ssh sut test -p /dev/sshserial`

Reply:
The tunnel-terminal tends to swallow characters while the tunnel is in place, every command in this terminal is very unreliable. This will not work.

Reply:
I don't think that this is the case here. The previous SSH command is way longer and works reliable.

Reply:
Maybe your experience initiates in your verification runs without `-y` where the previous SSH was messing your console up?

Reply:
No this is on the level of the terminal, see e.g. https://duck-norris.qam.suse.de/tests/10956#step/ssh_interactive_end/2 or https://duck-norris.qam.suse.de/tests/10956#step/ssh_interactive_end/2

While the command is active, the terminal swallows characters sometimes. Not the ssh tunnel, the serial terminal.

Reply:
But why? Maybe `-f` could help? But also because of the `while` loop you basically cannot kill that `ssh` background process because of it will be started again in 5 seconds.

Reply:
I asked and the response was basically "this is somewhat of a known issue and we don't know yet what to do about it". Preliminary investigations didn't gave me any results, my hypothesis is that the serial console doesn't like the piping stuff into `/dev/ttyS0` somehow but I don't have found the real cause nor a way to avoid this. And I really tried.

Reply:
again `ssh -q -o BatchMode=yes sut`?

Reply:
@b10n1k that would not help us here at all. We need the ssh tunnel itself, not a script. Also we know already that the instance is up and ready, that is handled elsewhere.

I don't see how the ssh BatchMode could help here in any way.